controller/node_controller.go

```diff
@@ -0,0 +1,228 @@
+		pStoreSynced: podInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "longhorn-node"),
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			n := obj.(*longhorn.Node)
+			nc.enqueueNode(n)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			cur := newObj.(*longhorn.Node)
+			nc.enqueueNode(cur)
+		},
+		DeleteFunc: func(obj interface{}) {
+			n := obj.(*longhorn.Node)
+			nc.enqueueNode(n)
+		},
+	})
+
+	return nc
+}
+
+func (nc *NodeController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer nc.queue.ShutDown()
+
+	logrus.Infof("Start Longhorn node controller")
+	defer logrus.Infof("Shutting down Longhorn node controller")
+
+	if !controller.WaitForCacheSync("longhorn node", stopCh, nc.pStoreSynced, nc.nStoreSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(nc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (nc *NodeController) worker() {
+	for nc.processNextWorkItem() {
+	}
+}
+
+func (nc *NodeController) processNextWorkItem() bool {
+	key, quit := nc.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer nc.queue.Done(key)
+
+	err := nc.syncNode(key.(string))
+	nc.handleErr(err, key)
+
+	return true
+}
+
+func (nc *NodeController) handleErr(err error, key interface{}) {
+	if err == nil {
+		nc.queue.Forget(key)
+		return
+	}
+
+	if nc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("Error syncing Longhorn node %v: %v", key, err)
+		nc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Longhorn node %v out of the queue: %v", key, err)
+	nc.queue.Forget(key)
+}
+
+func (nc *NodeController) syncNode(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "fail to sync node for %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != nc.namespace {
+		// Not ours, don't do anything
+		return nil
+	}
+
+	node, err := nc.ds.GetNode(name)
+	if err != nil {
+		return err
+	}
+	if node == nil {
+		logrus.Debugf("Longhorn node %v does not exist, regenerate a default one", key)
+		_, err = nc.ds.CreateDefaultNode(name)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if node.DeletionTimestamp != nil {
+		logrus.Infof("Remove longhorn node %v", node.Name)
+		return nc.ds.RemoveFinalizerForNode(node)
+	}
+
+	defer func() {
+		if err == nil {
+			_, err = nc.ds.UpdateNode(node)
+		}
+		// requeue if it's conflict
+		if apierrors.IsConflict(errors.Cause(err)) {
+			logrus.Debugf("Requeue %v due to conflict", key)
+			nc.enqueueNode(node)
+			err = nil
+		}
+	}()
+
+	// sync node state by manager pod
+	managerPods, err := nc.ds.ListManagerPods()
+	if err != nil {
+		return err
+	}
+	for _, pod := range managerPods {
+		err = nc.syncStatusWithPod(pod, node)
+		if err != nil {
+			return err
```

Comment:
As I said in the design doc, don't retry if conflict happens, assuming other managers are updating it. I am afraid the requeue may cause a storm of updating conflict in the larger system.